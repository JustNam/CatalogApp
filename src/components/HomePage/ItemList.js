/* eslint-disable arrow-parens */
import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { Pagination } from 'react-bootstrap';
import CreateItemModal from '../Modals/CreateItemModal';
import SuccessModal from '../Modals/SuccessModal';
import { historyWithRefresh } from '../../history';

class ItemList extends Component {
  constructor(props) {
    super(props);
    this.state = {
      showInfoModal: false,
      showSuccessModal: false,
    };
  }

  componentDidMount() {
    this.props.getCategories();
  }

  handleInfoShow = () => {
    this.setState({ showInfoModal: true });
  };

  handleInfoClose = () => {
    this.setState({ showInfoModal: false });
  };

  handleSuccessShow = () => {
    this.setState({ showSuccessModal: true });
  };

  handleSuccessClose = () => {
    this.setState({ showSuccessModal: false });
    historyWithRefresh.push('/categories');
  };

  getNewPage = (page) => {
    const { item } = this.props;
    this.props.getItemsInCategoryWithPagination(item.categoryId, page);
  }

  createPagination = () => {
    const { item } = this.props;
    const paginations = [];
    [...Array(item.lastPage).keys()].forEach((index) => {
      const page = index + 1;
      if (page === item.currentPage) {
        paginations.push(
          <Pagination.Item id={page} key={page} active>
            {page}
          </Pagination.Item>
        );
      } else {
        paginations.push(
          <Pagination.Item
            id={page}
            onClick={() => this.getNewPage(page)}
            key={page}
          >
            {page}
          </Pagination.Item>
        );
      }
    });
    return <Pagination>{paginations}</Pagination>;
  }

  renderItems(currentCategory, item) {
    const { showInfoModal, showSuccessModal } = this.state;
    return (
      <div>
        <div>
          <div className="list-header">
            {currentCategory[0] && `category ${currentCategory[0].name}`}
            <div className="pull-xs-right">
              <button
                id="createButton"
                className="btn btn-sm btn-outline-primary"
                type="button"
                onClick={this.handleInfoShow}
              >
                Create
              </button>
            </div>
          </div>
        </div>
        {item.data.map((itemDetail) => (
          <div className="article-preview" key={itemDetail.id}>
            <Link
              className="preview-link"
              to={`/categories/${item.categoryId}/items/${itemDetail.id}`}
            >
              <h1>{itemDetail.title}</h1>
              {(itemDetail.description && (
              <p>{(itemDetail.description.length > 100) ? `${itemDetail.description.substring(0, 100)}...` : itemDetail.description}</p>)
              )}
              <span>Read more</span>
            </Link>
          </div>
        ))}
        {this.createPagination()}
        <CreateItemModal
          currentCategoryId={currentCategory[0].id}
          show={showInfoModal}
          handleClose={this.handleInfoClose}
          showSuccessModal={this.handleSuccessShow}
        />
        <SuccessModal
          show={showSuccessModal}
          handleClose={this.handleSuccessClose}
        />
      </div>
    );
  }

  render() {
    const { item, category } = this.props;
    const currentCategory = category.data.filter(
      (categoryDetail) => parseInt(item.categoryId) === parseInt(categoryDetail.id)
    );
    if (item.data.length !== 0) {
      return this.renderItems(currentCategory, item);
    }
    return (
      <div>
        <div className="list-header">
          {currentCategory[0] && `category ${currentCategory[0].name}`}
          <div className="pull-xs-right">
            <button
              id="createButton"
              className="btn btn-sm btn-outline-primary"
              type="button"
              onClick={this.handleInfoShow}
            >
              Create
            </button>
          </div>
        </div>
        <div className="article-preview">
          <a className="preview-link">
            <i>
              <span>
                This category is empty. Please select another category!
              </span>
            </i>
          </a>
        </div>
        {currentCategory[0]
        && (
        <CreateItemModal
          currentCategoryId={currentCategory[0].id}
          show={this.state.showInfoModal}
          handleClose={this.handleInfoClose}
          showSuccessModal={this.handleSuccessShow}
        />
        )
        }
      </div>
    );
  }
}

export default ItemList;
