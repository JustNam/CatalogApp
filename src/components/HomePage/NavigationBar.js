import React, { Component } from 'react';
import 'styles/login.css';
import { connect } from 'react-redux';
import { NavDropdown } from 'react-bootstrap';
import { getItemsInCategoryWithPagination } from 'actions/item';
import { withRouter, Link } from 'react-router-dom';
import { logOut } from 'actions/user';

export class NavigationBar extends Component {
   reload = () => {
     const { category } = this.props;
     let id = parseInt(localStorage.getItem('categoryId'));
     if (!id) {
       id = category.data[0].id;
     }
     this.props.getItemsInCategoryWithPagination(id, 1);
   }

   render() {
     const username = localStorage.getItem('username');
     return (
       <nav className="navbar navbar-expand-sm bg-dark navbar-dark">
         <ul className="navbar-nav">
           <li className="nav-item active">
             <h1>
               <Link className="navbar-brand" onClick={this.reload} to="/categories">
                 Catalog App
               </Link>
             </h1>
           </li>
         </ul>
         <ul className="navbar-nav ml-auto">
           <NavDropdown title={username} id="basic-nav-dropdown">
             <NavDropdown.Divider />
             <NavDropdown.Item onClick={this.props.logOut}>Log out</NavDropdown.Item>
           </NavDropdown>
         </ul>
       </nav>
     );
   }
}
function mapStateToProps(state) {
  return {
    category: state.category,
  };
}
const mapDispatchToProp = {
  getItemsInCategoryWithPagination,
  logOut,
};

export default withRouter(connect(mapStateToProps, mapDispatchToProp)(NavigationBar));