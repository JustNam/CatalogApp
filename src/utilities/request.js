export const callAPI = async (enpoint) => {
  const response = await fetch(enpoint);
  const json = await response.json();
  return json;
};
