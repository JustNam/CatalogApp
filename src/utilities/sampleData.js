export const errorItem = {
  data: [
    {
      category: {
        created_on: '06/04/2019, 16:16:33',
        id: 1,
        name: 'Running',
        updated_on: '06/04/2019, 16:29:01',
      },
      created_on: '06/04/2019, 16:29:43',
      id: 7,
      title: 'Ru4)',
      updated_on: '07/02/2019, 22:08:02',
      user: {
        id: 1,
        username: 'nam123',
      },
    },
  ],
};

export const itemWithEmptyCategory = {
  categoryId: 1,
  currentPage: 1,
  lastPage: 1,
  data: [],
};

export const noDescriptionItem = {
  categoryId: 1,
  currentPage: 1,
  lastPage: 1,
  data: [
    {
      category: {
        created_on: '06/04/2019, 16:16:33',
        id: 1,
        name: 'Running',
        updated_on: '06/04/2019, 16:29:01',
      },
      created_on: '06/04/2019, 16:29:43',
      id: 7,
      title: 'Ru4',
      updated_on: '07/02/2019, 22:08:02',
      user: {
        id: 1,
        username: 'nam123',
      },
    },
  ],
};

export const item = {
  categoryId: 1,
  currentPage: 1,
  lastPage: 3,
  data: [
    {
      category: {
        created_on: '06/04/2019, 16:16:33',
        id: 1,
        name: 'Running',
        updated_on: '06/04/2019, 16:29:01',
      },
      created_on: '06/04/2019, 16:29:43',
      description: 'Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry\'s standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum',
      id: 7,
      title: 'Ru4',
      updated_on: '07/02/2019, 22:08:02',
      user: {
        id: 1,
        username: 'nam123',
      },
    },
    {
      category: {
        created_on: '06/04/2019, 16:16:33',
        id: 1,
        name: 'Running',
        updated_on: '06/04/2019, 16:29:01',
      },
      created_on: '06/04/2019, 16:29:43',
      id: 8,
      title: 'Ru3',
      updated_on: '07/02/2019, 22:08:02',
      user: {
        id: 1,
        username: 'nam123',
      },
    },
    {
      category: {
        created_on: '06/04/2019, 16:16:33',
        id: 1,
        name: 'Running',
        updated_on: '06/04/2019, 16:29:01',
      },
      created_on: '06/04/2019, 16:29:43',
      description: 'Lorem Ipsum is simply',
      id: 9,
      title: 'Ru2',
      updated_on: '07/02/2019, 22:08:02',
      user: {
        id: 1,
        username: 'nam123',
      },
    },
  ],
};

export const category = {
  data: [
    {
      created_on: '06/26/2019, 08:09:32',
      id: 1,
      name: 'Swimming',
      updated_on: '06/26/2019, 08:09:32',
    },
    {
      created_on: '06/26/2019, 08:09:32',
      id: 2,
      name: 'Running',
      updated_on: '06/26/2019, 08:09:32',
    },
  ],
};
