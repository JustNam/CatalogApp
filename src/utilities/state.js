export const initalCategoryState = {
  data: [],
};

export const initalItemState = {
  data: [],
  categoryId: 0,
  currentPage: 0,
  lastPage: 0,
};

export const initalUserState = {
  loggedIn: localStorage.getItem('accessToken') || false,
};