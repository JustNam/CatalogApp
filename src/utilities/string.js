export const convertCamelCaseToSnakeCase = string => (
  string.replace(/[A-Z]/g, match => `_${match.toLowerCase()}`)
);