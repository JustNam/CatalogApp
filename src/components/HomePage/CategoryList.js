import React, { Component } from 'react';
import { Tabs, Tab, TabPanel, TabList } from 'react-web-tabs';
import 'react-web-tabs/dist/react-web-tabs.css';
import ItemList from 'components/HomePage/ItemList';
import NavigationBar from 'components/HomePage/NavigationBar';


class CategoryList extends Component {
  constructor(props) {
    super(props);
    this.state = {
      categoryId: 0,
    };
  }

  async componentDidMount() {
    // Load the data of default category
    await this.props.getCategories().then(() => {
      const { category } = this.props;
      const { id } = category.data[0];
      this.props.getItemsInCategoryWithPagination(id, 1);
      this.setState({ categoryId: id });
    });
  }

  getItems(e, categoryId) {
    // Get items in chosen category
    e.preventDefault();
    this.props.getItemsInCategoryWithPagination(categoryId, 1);
    this.setState({ categoryId });
  }

  render() {
    // category variable contains the general information about categories
    const { category } = this.props;
    const { categoryId } = this.state;
    return (
      <div>
        <NavigationBar />
        <div className="container page">
          <Tabs defaultTab="vertical-tab-0" vertical>
            <TabList>
              {category.data.map((category, index) => (
                <Tab
                  tabFor={`vertical-tab-${index}`}
                  key={category.id}
                  onClick={e => this.getItems(e, category.id)}
                >
                  <h3>{category.name}</h3>
                </Tab>
              ))}
              <br />
            </TabList>
            {category.data.map((category, index) => (
              <TabPanel tabId={`vertical-tab-${index}`} key={category.id}>
                {category.id === categoryId && <ItemList {...this.props} />}
              </TabPanel>
            ))}
          </Tabs>
        </div>
      </div>
    );
  }
}

export default CategoryList;
