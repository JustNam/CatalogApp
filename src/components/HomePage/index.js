import CategoryList from 'components/HomePage/CategoryList';

export default CategoryList;