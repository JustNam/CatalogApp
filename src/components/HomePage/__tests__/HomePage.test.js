import React from 'react';
import { shallow } from 'enzyme';
import { HomePage } from 'components/HomePage/HomePage';
import { category } from 'utilities/sampleData';

describe('/components/HomePage', () => {
  let wrapper;
  let props;
  const setup = () => {
    props = {
      category,
      getCategories: jest.fn(() => Promise.resolve()),
      getItems: jest.fn(() => Promise.resolve()),
      getItemsInCategoryWithPagination: jest.fn(() => Promise.resolve()),
    };
  };
  beforeEach(() => {
    setup();
    wrapper = shallow(<HomePage {...props} />);
  });

  it('It should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
  it('It should get the items in first category', () => {
    expect(props.getItemsInCategoryWithPagination).toBeCalled();
  });
  it('It should get the items in another category when another tab is clicked', () => {
    wrapper.find('[tabFor="vertical-tab-1"]').simulate('click', {
      preventDefault: () => {},
    });
    expect(props.getItemsInCategoryWithPagination.mock.calls.length).toBe(2);
  });
});
