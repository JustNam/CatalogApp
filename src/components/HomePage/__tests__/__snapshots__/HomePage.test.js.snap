// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/components/HomePage It should render correctly 1`] = `ShallowWrapper {}`;
